File: src/algorithms/genetic_algorithm.cpp
// filepath: /Users/punam/Desktop/Varsity/2-2/Algo/project/github deployment/ConflictFreeScheduling/src/algorithms/genetic_algorithm.cpp
#include "genetic_algorithm.h"
#include <random>
#include <algorithm>
#include <numeric>
#include <iostream>

std::vector<Activity> GeneticAlgorithmScheduler::evolveSchedule(std::vector<Activity> activities, const GAConfig& config) {
    if (activities.empty()) return {};
    
    std::mt19937 rng(std::random_device{}());
    
    // Initialize population
    auto population = initializePopulation(config.populationSize, activities.size(), rng);
    
    // Evaluate initial population
    for (auto& individual : population) {
        individual.fitness = calculateFitness(individual, activities);
    }
    
    // Evolution loop
    for (size_t generation = 0; generation < config.generations; ++generation) {
        // Sort by fitness (descending)
        std::sort(population.begin(), population.end(), 
                  [](const Individual& a, const Individual& b) { 
                      return a.fitness > b.fitness; 
                  });
        
        // Create new generation
        std::vector<Individual> newPopulation;
        
        // Keep elite individuals
        for (size_t i = 0; i < config.eliteSize && i < population.size(); ++i) {
            newPopulation.push_back(population[i]);
        }
        
        // Generate offspring
        while (newPopulation.size() < config.populationSize) {
            auto parents = selectParents(population, rng);
            auto offspring = crossover(parents.first, parents.second, config.crossoverRate, rng);
            
            mutate(offspring.first, config.mutationRate, rng);
            mutate(offspring.second, config.mutationRate, rng);
            
            offspring.first.fitness = calculateFitness(offspring.first, activities);
            offspring.second.fitness = calculateFitness(offspring.second, activities);
            
            newPopulation.push_back(offspring.first);
            if (newPopulation.size() < config.populationSize) {
                newPopulation.push_back(offspring.second);
            }
        }
        
        population = std::move(newPopulation);
        
        // Optional: Print progress every 100 generations
        if (generation % 100 == 0) {
            std::cout << "Generation " << generation << ", Best fitness: " << population[0].fitness << std::endl;
        }
    }
    
    // Return best solution
    std::sort(population.begin(), population.end(), 
              [](const Individual& a, const Individual& b) { 
                  return a.fitness > b.fitness; 
              });
    
    return individualToActivities(population[0], activities);
}

std::vector<GeneticAlgorithmScheduler::Individual> GeneticAlgorithmScheduler::initializePopulation(
    size_t populationSize, size_t numActivities, std::mt19937& rng) {
    
    std::vector<Individual> population;
    std::uniform_real_distribution<double> prob(0.0, 1.0);
    
    for (size_t i = 0; i < populationSize; ++i) {
        Individual individual(numActivities);
        
        // Random initialization with 30% probability for each gene
        for (size_t j = 0; j < individual.genes.size(); ++j) {
            individual.genes[j] = prob(rng) < 0.3;
        }
        
        population.push_back(individual);
    }
    
    return population;
}

double GeneticAlgorithmScheduler::calculateFitness(Individual& individual, const std::vector<Activity>& activities) {
    auto schedule = individualToActivities(individual, activities);
    
    if (schedule.empty()) return 0.0;
    
    // Calculate total weight
    double totalWeight = 0.0;
    for (const auto& activity : schedule) {
        totalWeight += activity.weight;
    }
    
    // Calculate conflict penalty
    double conflictPenalty = calculateConflictPenalty(schedule);
    
    // Fitness = total weight - heavy penalty for conflicts
    return totalWeight - (conflictPenalty * 1000.0);
}

std::pair<GeneticAlgorithmScheduler::Individual, GeneticAlgorithmScheduler::Individual> 
GeneticAlgorithmScheduler::selectParents(const std::vector<Individual>& population, std::mt19937& rng) {
    // Tournament selection
    std::uniform_int_distribution<size_t> dist(0, population.size() - 1);
    
    auto tournamentSelect = [&]() -> const Individual& {
        const size_t tournamentSize = 3;
        size_t bestIdx = dist(rng);
        double bestFitness = population[bestIdx].fitness;
        
        for (size_t i = 1; i < tournamentSize; ++i) {
            size_t idx = dist(rng);
            if (population[idx].fitness > bestFitness) {
                bestFitness = population[idx].fitness;
                bestIdx = idx;
            }
        }
        return population[bestIdx];
    };
    
    return {tournamentSelect(), tournamentSelect()};
}

std::pair<GeneticAlgorithmScheduler::Individual, GeneticAlgorithmScheduler::Individual> 
GeneticAlgorithmScheduler::crossover(const Individual& parent1, const Individual& parent2, 
                                    double crossoverRate, std::mt19937& rng) {
    Individual child1 = parent1;
    Individual child2 = parent2;
    
    std::uniform_real_distribution<double> prob(0.0, 1.0);
    
    if (prob(rng) < crossoverRate) {
        // Single-point crossover
        std::uniform_int_distribution<size_t> pointDist(1, parent1.genes.size() - 1);
        size_t crossoverPoint = pointDist(rng);
        
        for (size_t i = crossoverPoint; i < parent1.genes.size(); ++i) {
            child1.genes[i] = parent2.genes[i];
            child2.genes[i] = parent1.genes[i];
        }
    }
    
    return {child1, child2};
}

void GeneticAlgorithmScheduler::mutate(Individual& individual, double mutationRate, std::mt19937& rng) {
    std::uniform_real_distribution<double> prob(0.0, 1.0);
    
    for (size_t i = 0; i < individual.genes.size(); ++i) {
        if (prob(rng) < mutationRate) {
            individual.genes[i] = !individual.genes[i];
        }
    }
}

std::vector<Activity> GeneticAlgorithmScheduler::individualToActivities(
    const Individual& individual, const std::vector<Activity>& activities) {
    
    std::vector<Activity> result;
    for (size_t i = 0; i < individual.genes.size(); ++i) {
        if (individual.genes[i]) {
            result.push_back(activities[i]);
        }
    }
    return result;
}

bool GeneticAlgorithmScheduler::hasConflict(const Activity& a1, const Activity& a2) {
    return !(a1.end <= a2.start || a2.end <= a1.start);
}

double GeneticAlgorithmScheduler::calculateConflictPenalty(const std::vector<Activity>& schedule) {
    double penalty = 0.0;
    
    for (size_t i = 0; i < schedule.size(); ++i) {
        for (size_t j = i + 1; j < schedule.size(); ++j) {
            if (hasConflict(schedule[i], schedule[j])) {
                penalty += 1.0;
            }
        }
    }
    
    return penalty;
}

File: src/algorithms/genetic_algorithm.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Genetic Algorithm for Conflict-Free Scheduling

This module implements a genetic algorithm for evolutionary optimization
of activity scheduling problems.
"""

from typing import List, Tuple
from dataclasses import dataclass
import random
import copy


@dataclass
class Activity:
    """Class representing an activity/task with start and end times"""
    id: int
    start: int
    end: int
    weight: float = 1.0  # Default weight for unweighted problems
    name: str = ""
    room: str = ""


@dataclass
class GAConfig:
    """Configuration for Genetic Algorithm"""
    population_size: int = 50
    generations: int = 100
    crossover_rate: float = 0.8
    mutation_rate: float = 0.1
    elite_size: int = 5


class Individual:
    """Represents an individual in the genetic algorithm population"""
    
    def __init__(self, activities: List[Activity], chromosome: List[bool] = None):
        """
        Initialize individual
        
        Args:
            activities: List of all activities
            chromosome: Binary representation (True = selected, False = not selected)
        """
        self.activities = activities
        if chromosome is None:
            # Random initialization
            self.chromosome = [random.choice([True, False]) for _ in activities]
        else:
            self.chromosome = chromosome
        
        self.fitness = 0.0
        self.is_valid = False
        self._evaluate()
    
    def _evaluate(self) -> None:
        """Evaluate fitness of this individual"""
        selected_activities = []
        for i, selected in enumerate(self.chromosome):
            if selected:
                selected_activities.append(self.activities[i])
        
        # Check if the selection is conflict-free
        self.is_valid = self._is_conflict_free(selected_activities)
        
        if self.is_valid:
            # Fitness is total weight of selected activities
            self.fitness = sum(act.weight for act in selected_activities)
        else:
            # Penalty for invalid solutions
            total_weight = sum(act.weight for act in selected_activities)
            conflicts = self._count_conflicts(selected_activities)
            self.fitness = total_weight - (conflicts * 10)  # Heavy penalty for conflicts
    
    def _is_conflict_free(self, activities: List[Activity]) -> bool:
        """Check if selected activities are conflict-free"""
        if len(activities) <= 1:
            return True
        
        # Sort by start time
        activities = sorted(activities, key=lambda a: a.start)
        
        for i in range(len(activities) - 1):
            if activities[i].end > activities[i + 1].start:
                return False
        
        return True
    
    def _count_conflicts(self, activities: List[Activity]) -> int:
        """Count number of conflicts in the selection"""
        conflicts = 0
        n = len(activities)
        
        for i in range(n):
            for j in range(i + 1, n):
                if not (activities[i].end <= activities[j].start or 
                       activities[j].end <= activities[i].start):
                    conflicts += 1
        
        return conflicts
    
    def get_selected_activities(self) -> List[Activity]:
        """Get list of selected activities"""
        return [self.activities[i] for i, selected in enumerate(self.chromosome) if selected]


class GeneticAlgorithmScheduler:
    """Genetic Algorithm for Evolutionary Optimization"""
    
    @staticmethod
    def evolve_schedule(activities: List[Activity], config: GAConfig) -> List[Activity]:
        """
        Evolve optimal schedule using genetic algorithm
        
        Args:
            activities: List of activities to schedule
            config: GA configuration parameters
            
        Returns:
            List of activities in evolved optimal schedule
        """
        if not activities:
            return []
        
        # Initialize population
        population = []
        for _ in range(config.population_size):
            individual = Individual(activities)
            population.append(individual)
        
        best_individual = None
        best_fitness = float('-inf')
        
        # Evolution loop
        for generation in range(config.generations):
            # Evaluate population
            for individual in population:
                if individual.fitness > best_fitness and individual.is_valid:
                    best_fitness = individual.fitness
                    best_individual = copy.deepcopy(individual)
            
            # Selection, crossover, and mutation
            new_population = []
            
            # Elite selection (keep best individuals)
            population.sort(key=lambda x: x.fitness, reverse=True)
            for i in range(config.elite_size):
                if i < len(population):
                    new_population.append(copy.deepcopy(population[i]))
            
            # Generate rest of population through crossover and mutation
            while len(new_population) < config.population_size:
                # Selection
                parent1 = GeneticAlgorithmScheduler._tournament_selection(population)
                parent2 = GeneticAlgorithmScheduler._tournament_selection(population)
                
                # Crossover
                if random.random() < config.crossover_rate:
                    child1, child2 = GeneticAlgorithmScheduler._crossover(parent1, parent2, activities)
                else:
                    child1, child2 = copy.deepcopy(parent1), copy.deepcopy(parent2)
                
                # Mutation
                if random.random() < config.mutation_rate:
                    GeneticAlgorithmScheduler._mutate(child1)
                if random.random() < config.mutation_rate:
                    GeneticAlgorithmScheduler._mutate(child2)
                
                new_population.extend([child1, child2])
            
            # Trim population to exact size
            population = new_population[:config.population_size]
        
        # Return best solution found
        if best_individual and best_individual.is_valid:
            return best_individual.get_selected_activities()
        else:
            # Fallback: return greedy solution
            return GeneticAlgorithmScheduler._greedy_fallback(activities)
    
    @staticmethod
    def _tournament_selection(population: List[Individual], tournament_size: int = 3) -> Individual:
        """
        Tournament selection for parent selection
        
        Args:
            population: Current population
            tournament_size: Size of tournament
            
        Returns:
            Selected individual
        """
        tournament = random.sample(population, min(tournament_size, len(population)))
        tournament.sort(key=lambda x: x.fitness, reverse=True)
        return tournament[0]
    
    @staticmethod
    def _crossover(parent1: Individual, parent2: Individual, 
                  activities: List[Activity]) -> Tuple[Individual, Individual]:
        """
        Single-point crossover
        
        Args:
            parent1: First parent
            parent2: Second parent
            activities: List of activities
            
        Returns:
            Tuple of two children
        """
        crossover_point = random.randint(1, len(parent1.chromosome) - 1)
        
        child1_chromosome = (parent1.chromosome[:crossover_point] + 
                           parent2.chromosome[crossover_point:])
        child2_chromosome = (parent2.chromosome[:crossover_point] + 
                           parent1.chromosome[crossover_point:])
        
        child1 = Individual(activities, child1_chromosome)
        child2 = Individual(activities, child2_chromosome)
        
        return child1, child2
    
    @staticmethod
    def _mutate(individual: Individual) -> None:
        """
        Bit-flip mutation
        
        Args:
            individual: Individual to mutate
        """
        mutation_point = random.randint(0, len(individual.chromosome) - 1)
        individual.chromosome[mutation_point] = not individual.chromosome[mutation_point]
        individual._evaluate()  # Re-evaluate after mutation
    
    @staticmethod
    def _greedy_fallback(activities: List[Activity]) -> List[Activity]:
        """
        Greedy fallback solution when GA fails
        
        Args:
            activities: List of activities
            
        Returns:
            Greedy solution
        """
        if not activities:
            return []
        
        # Sort by weight/duration ratio for better greedy selection
        activities = sorted(activities, key=lambda a: a.weight / max(a.end - a.start, 1), reverse=True)
        
        selected = []
        for activity in activities:
            # Check if this activity conflicts with any selected activity
            conflicts = False
            for selected_activity in selected:
                if not (activity.end <= selected_activity.start or 
                       selected_activity.end <= activity.start):
                    conflicts = True
                    break
            
            if not conflicts:
                selected.append(activity)
        
        return selected
    
    @staticmethod
    def evolve_with_multiple_runs(activities: List[Activity], config: GAConfig, 
                                runs: int = 5) -> List[Activity]:
        """
        Run GA multiple times and return best result
        
        Args:
            activities: List of activities to schedule
            config: GA configuration
            runs: Number of independent runs
            
        Returns:
            Best schedule found across all runs
        """
        best_schedule = []
        best_weight = 0.0
        
        for _ in range(runs):
            schedule = GeneticAlgorithmScheduler.evolve_schedule(activities, config)
            weight = sum(act.weight for act in schedule)
            
            if weight > best_weight:
                best_weight = weight
                best_schedule = schedule
        
        return best_schedule
